rbac/providers/ldap/ldap_sync.py

```diff
@@ -0,0 +1,138 @@
+#! /usr/bin/env python3
+
+# Copyright 2018 Contributors to Hyperledger Sawtooth
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# -----------------------------------------------------------------------------
+
+# http://docs.python-requests.org/en/master/
+
+import json
+import os
+from datetime import datetime, timezone
+from threading import Timer
+
+import ldap3
+import rethinkdb as r
+from ldap3 import ALL, Connection, Server
+from rbac.providers.inbound_filters import (inbound_group_filter,
+                                            inbound_user_filter)
+
+DEFAULT_CONFIG = {
+    "DB_HOST": "rethink",
+    "DB_PORT": 28015,
+    "DB_NAME": "rbac",
+    "DELTA_SYNC_INTERVAL_SECONDS": 3600.0,
+}
+
+
+def getenv(name, default):
+    value = os.getenv(name)
+    if value is None or value is "":
+        return default
+    return value
+
+
+DB_HOST = getenv("DB_HOST", DEFAULT_CONFIG["DB_HOST"])
+DB_PORT = getenv("DB_PORT", DEFAULT_CONFIG["DB_PORT"])
+DB_NAME = getenv("DB_NAME", DEFAULT_CONFIG["DB_NAME"])
+
+LDAP_DC = os.environ.get("LDAP_DC")
+LDAP_SERVER = os.environ.get("LDAP_SERVER")
+LDAP_USER = os.environ.get("LDAP_USER")
+LDAP_PASS = os.environ.get("LDAP_PASS")
+DELTA_SYNC_INTERVAL_SECONDS = getenv("DELTA_SYNC_INTERVAL_SECONDS", DEFAULT_CONFIG["DELTA_SYNC_INTERVAL_SECONDS"])
+
+
+def fetch_ldap_data(sync_type, data_type):
+    """
+        Call to get entries for all (Users | Groups) in Active Directory, saves the time of the sync,
+        and inserts data into RethinkDB.
+    """
+
+    r.connect(host=DB_HOST, port=DB_PORT, db=DB_NAME).repl()
+
+    if sync_type == "delta":
+        last_sync = r.table("sync_tracker").filter({"source": "ldap-" + data_type}).coerce_to('array').run()
+        last_sync_time = datetime.strptime(last_sync[0]['timestamp'].split('+')[0], "%Y-%m-%dT%H:%M:%S.%f")
```

Comment:
Please separate data transformation from data sync. That is: Move the transformations into their own python transformer file. Doing so should make this block of code easier to read by abstracting the regex and string arithmetic behind functions having meaningful names, and it will make both the transforms and sync code easier to test.

Reply:
I created a new data transformer file for LDAP provider.